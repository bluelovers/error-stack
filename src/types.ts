import { ITSPartialRecord } from 'ts-type/lib/type/record';

export interface ISource
{
	/**
	 * The source of the the callee
	 */
	source: string
	line?: string
	col?: string
}

export interface ITrace extends ISource
{
	raw?: false

	callee: string
	calleeNote?: string
	/**
	 * Whether the callee is 'eval'
	 */
	eval?: boolean

	evalCallee?: string
	evalCalleeNote?: string

	/**
	 * The source location inside eval content
	 */
	evalTrace?: ISource

	indent?: string,
}

export interface IRawLineTrace extends ITSPartialRecord<Exclude<keyof ITrace, 'raw' | 'indent' | 'rawLine'>, undefined>
{
	raw: true
	indent?: string,
	rawLine: string,
}

export interface IEvalTrace extends ITrace
{
	eval: true
	callee: "eval",
	evalTrace: ISource
}

export type ITraceValue = IRawLineTrace | ITrace | IEvalTrace;

export interface IParsedWithoutTrace
{
	/**
	 * Error type
	 */
	type: string;
	code?: string;
	/**
	 * The message used by Error constructor
	 */
	message: string;
}

export interface IParsed extends IParsedWithoutTrace
{
	traces: ITraceValue[];

	rawMessage?: string,
	rawTrace?: string[],
	rawStack?: string,
}
