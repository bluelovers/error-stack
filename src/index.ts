import { ITSPickExtra, ITSRequireAtLeastOne } from 'ts-type/lib/type/record';
import { lineSplit, R_CRLF, LF } from 'crlf-normalize';
import { IEvalTrace, IParsed, IParsedWithoutTrace, ISource, ITrace, IRawLineTrace, ITraceValue } from './types';
import { stringSplitWithLimit } from 'string-split-keep2';
import { trim } from './util/trim';
import { isUnset } from './util/isUnset';
import { isNumOnly } from './util/isNumOnly';
import errcode from 'err-code';
import { inspect } from 'util';

const AT = 'at' as const

// 1.
// Error: foo
// 2.
// TypeError: foo
const REGEX_MATCH_MESSAGE = /^([a-z][a-z0-9_]*)(?:(?: \[(\w+)\])?:(?: ([\s\S]*))?)?$/i

const REGEX_MATCH_MESSAGE_LOOSE = new RegExp(REGEX_MATCH_MESSAGE.source, REGEX_MATCH_MESSAGE.flags + 'm');

const REGEX_REMOVE_AT = /^at\s+/
const REGEX_STARTS_WITH_EVAL_AT = /^eval\s+at\s+/

const REGEX_MATCH_INDENT = /^([ \t]*)(.+)$/;

export function breakBrackets(str: string, first: string, last: string)
{
	if (!str.endsWith(last))
	{
		return [str]
	}

	let firstIndex
	let cursor = str.length - 1
	// There is already the last one
	let count = 1
	while (--cursor >= 0)
	{
		const char = str.charAt(cursor)
		if (char === last)
		{
			count++
		}
		else if (char === first)
		{
			if (--count === 0)
			{
				firstIndex = cursor
				break
			}
		}
	}

	return [
		str.slice(0, firstIndex),
		str.slice(firstIndex + 1, -1),
	].map(trim)
}

export function validPosition(source: {
	line?: string | number,
	col?: string | number,
})
{
	if (!isUnset(source))
	{
		if (typeof source === 'object' && isUnset(source.line) && isUnset(source.col))
		{
			return null
		}

		return isNumOnly(source.line) && isNumOnly(source.col)
	}

	return false
}

export function parseSource(rawSource: string): ISource
{
	const [source, line, col] = stringSplitWithLimit(rawSource, ':', -3);

	if (!col?.length || !line?.length)
	{
		return {
			source: rawSource,
		}
	}

	return {
		source,
		line,
		col,
	}
}

export function parseEvalSource(rawEvalSource: string): Omit<IEvalTrace, 'callee' | 'calleeNote' | 'eval'>
{
	const { indent, rawLine } = _detectIndent(rawEvalSource);

	const [rawTrace, rawEvalTrace] = rawLine
		.replace(REGEX_STARTS_WITH_EVAL_AT, '')
		.split(/,\s+/g)
		.map(trim)

	const {
		eval: ev,
		callee: evalCallee,
		calleeNote: evalCalleeNote,
		...trace
	} = parseTrace(rawTrace)

	const evalTrace = parseSource(rawEvalTrace)

	return {
		evalCallee,
		evalCalleeNote,
		...trace,
		evalTrace,
		indent,
	}
}

export function _detectIndent(trace: string)
{
	const [, indent, rawLine] = REGEX_MATCH_INDENT.exec(trace)

	return {
		indent,
		rawLine,
	}
}

export function parseTrace(trace: string, testEvalSource: true): ITrace | IRawLineTrace
export function parseTrace(trace: string, testEvalSource?: false): ITrace
export function parseTrace(trace: string, testEvalSource?: boolean): ITrace | IRawLineTrace
export function parseTrace(trace: string, testEvalSource?: boolean)
{
	const { indent, rawLine } = _detectIndent(trace);

	const t = rawLine.replace(REGEX_REMOVE_AT, '')

	let [
		rawCallee, rawSource,
	] = breakBrackets(t, '(', ')')

	if (!rawSource)
	{
		[rawCallee, rawSource] = [rawSource, rawCallee]
	}

	const ret: ITrace = {} as any

	if (rawCallee)
	{
		const [
			callee, calleeNote,
		] = breakBrackets(rawCallee, '[', ']')

		ret.callee = callee
		ret.calleeNote = calleeNote
	}
	else
	{
		ret.callee = rawCallee
	}

	if (ret.callee === 'eval')
	{
		ret.eval = true
	}

	if (testEvalSource === true)
	{
		if (!rawLine.startsWith(AT))
		{
			return <IRawLineTrace>{
				raw: true,
				indent,
				rawLine,
			}
		}
	}

	Object.assign(
		ret,
		testEvalSource && isEvalSource(rawSource)
			? parseEvalSource(rawSource)
			: parseSource(rawSource),
	)

	if (testEvalSource === true)
	{
		if (!validTrace(ret))
		{
			return <IRawLineTrace>{
				raw: true,
				indent,
				rawLine,
			}
		}
	}

	ret.indent = indent

	return ret
}

export function isEvalSource(rawSource: string)
{
	return REGEX_STARTS_WITH_EVAL_AT.test(rawSource)
}

export function validTrace(trace: ITraceValue)
{
	if (isRawLineTrace(trace))
	{
		return false;
	}

	return trace.eval || isNumOnly(trace.line) || isUnset(trace.callee) && trace.source?.length > 0 && validPosition(trace);
}

export function parseBody(rawStack: string, detectMessage?: string)
{
	let rawTrace: string[];
	let rawMessage: string

	if (!isUnset(detectMessage))
	{
		let { type, message } = parseMessage(rawStack, true);

		let mf = formatMessage({
			type,
			message: detectMessage === '' ? message : detectMessage,
		});

		let i = rawStack.indexOf(mf)

		if (i === 0)
		{
			let s = rawStack.replace(mf, '')
			let m = R_CRLF.exec(s)

			if (m?.index === 0)
			{
				rawTrace = lineSplit(m.input.replace(m[0], ''));

				rawMessage = mf
			}
		}
	}

	if (!rawMessage?.length)
	{
		([rawMessage, ...rawTrace] = lineSplit(rawStack));

		// A error message might have multiple lines
		const index = rawTrace.findIndex(line => line.trimLeft()
			.startsWith(AT) && validTrace(parseTrace(trim(line), true)));

		rawMessage = [rawMessage, ...rawTrace.splice(0, index)].join(LF)
	}

	return {
		rawMessage,
		rawTrace,
	}
}

export function parseMessage(body: string, looseMode?: boolean): IParsedWithoutTrace
{
	try
	{
		const [, type, code, message] = body.match(looseMode ? REGEX_MATCH_MESSAGE_LOOSE : REGEX_MATCH_MESSAGE);

		return {
			type,
			code,
			message,
		}
	}
	catch (e: any)
	{
		e.message = `Failed to parse error message.\nreason: ${e.message}\nbody=${inspect(body)}`;

		errcode(e, {
			body,
		});

		throw e
	}
}

export function parseStack(rawStack: string, detectMessage?: string): IParsed
{
	if (typeof rawStack !== 'string')
	{
		throw errcode(new TypeError('stack must be a string'), {
			rawStack,
			detectMessage,
		});
	}

	try
	{
		const { rawMessage, rawTrace } = parseBody(rawStack, detectMessage);

		const {
			type, code, message,
		} = parseMessage(rawMessage)

		const traces = rawTrace.map(t => parseTrace(t, true))

		return {
			type,
			code,
			message,
			traces,
			rawMessage,
			rawTrace,
			rawStack,
		}
	}
	catch (e: any)
	{
		errcode(e, {
			rawStack,
			detectMessage,
		});

		throw e;
	}
}

export function formatTrace({
	callee,
	calleeNote,
	source,
	line,
	col,
}: ITSPickExtra<ITrace, 'source'>)
{
	const sourceTrace = [
		source,
		line,
		col,
	]
		.filter(v => typeof v !== 'undefined')
		.join(':')

	const note = calleeNote
		? ` [${calleeNote}]`
		: ''

	return callee
		? `${callee}${note} (${sourceTrace})`
		: sourceTrace
}

export function formatEvalTrace({
	callee,
	evalTrace,

	evalCallee,
	evalCalleeNote,

	...trace
}: IEvalTrace)
{
	return `${callee} (eval at ${formatTrace({
		...trace,

		callee: evalCallee ?? '<anonymous>',
		calleeNote: evalCalleeNote,
	})}, ${formatTrace(evalTrace)})`;
}

export function formatMessagePrefix({
	type,
	code,
}: IParsedWithoutTrace)
{
	if (code?.length)
	{
		type += ` [${code}]`;
	}

	return `${type}`;
}

export function formatMessage(parsed: IParsedWithoutTrace)
{
	let line = formatMessagePrefix(parsed);

	if (typeof parsed.message !== 'undefined')
	{
		line += `: ${parsed.message ?? ''}`;
	}

	return line;
}

export function formatRawLineTrace(trace: IRawLineTrace)
{
	return `${trace.indent ?? '    '}${trace.rawLine}`
}

export function isRawLineTrace(trace: ITraceValue): trace is IRawLineTrace
{
	return (trace.raw === true)
}

export function isEvalTrace(trace: ITraceValue): trace is IEvalTrace
{
	return (trace as IEvalTrace).eval === true
}

export function formatTraceLine(trace: ITraceValue)
{
	if (isRawLineTrace(trace))
	{
		return formatRawLineTrace(trace)
	}

	return `${trace.indent ?? '    '}at ${
		isEvalTrace(trace)
			? formatEvalTrace(trace)
			: formatTrace(trace)
	}`
}

export class ErrorStack implements IParsed
{

	/**
	 * Error type
	 */
	type: string;

	code?: string;
	/**
	 * The message used by Error constructor
	 */
	message: string;
	traces: IParsed["traces"];

	readonly rawMessage?: string;
	readonly rawTrace?: string[];
	readonly rawStack?: string;

	constructor(stack: string, detectMessage?: string)
	{
		Object.assign(this, parseStack(stack, detectMessage))
	}

	/**
	 * filterFunction Function the same as the callback function of Array.prototype.filter(callback)
	 */
	filter(filter: (value: ITraceValue, index: number, array: IParsed["traces"]) => boolean)
	{
		this.traces = this.traces.filter(filter)

		return this
	}

	/**
	 * Format object parsed
	 */
	format()
	{
		return stringifyErrorStack(this)
	}
}

export function formatTraces(traces: IParsed["traces"])
{
	return traces?.map(formatTraceLine)
}

/**
 * Format object parsed
 */
export function stringifyErrorStack(parsed: ITSRequireAtLeastOne<IParsed, 'traces' | 'rawTrace'>)
{
	const messageLines = `${formatMessage(parsed)}`
	const tracesLines = (parsed.traces?.map(formatTraceLine) ?? parsed.rawTrace)
		.join(LF)
	;

	return tracesLines
		? messageLines + LF + tracesLines
		: messageLines
}

export function parseErrorStack(stack: string, detectMessage?: string)
{
	return new ErrorStack(stack, detectMessage)
}

export default parseErrorStack
