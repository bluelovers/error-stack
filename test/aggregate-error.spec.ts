import { parseErrorStack, stringifyErrorStack } from '../src/index';

[
	[
		'AggregateError', 'AggregateError\n' +
	'    at Object.<anonymous> (G:\\Users\\The Project\\fork\\error-stack\\test\\temp2.ts:4:11)\n' +
	'    at Module._compile (node:internal/modules/cjs/loader:1097:14)\n' +
	'    at Module.m._compile (C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\ts-node\\src\\index.ts:1371:23)\n' +
	'    at Module._extensions..js (node:internal/modules/cjs/loader:1149:10)\n' +
	'    at Object.require.extensions.<computed> [as .ts] (C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\ts-node\\src\\index.ts:1374:12)\n' +
	'    at Module.load (node:internal/modules/cjs/loader:975:32)\n' +
	'    at Function.Module._load (node:internal/modules/cjs/loader:822:12)\n' +
	'    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)\n' +
	'    at main (C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\ts-node\\src\\bin.ts:331:12)\n' +
	'    at Object.<anonymous> (C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\ts-node\\src\\bin.ts:482:3)',
	],
	[
		'AggregateError: ', 'AggregateError: \n' +
	'    at Object.<anonymous> (G:\\Users\\The Project\\nodejs-yarn\\ws-error\\packages\\lazy-aggregate-error\\test\\temp2.ts:3:9)\n' +
	'    at Module._compile (node:internal/modules/cjs/loader:1097:14)\n' +
	'    at Module.m._compile (C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\ts-node\\src\\index.ts:1371:23)\n' +
	'    at Module._extensions..js (node:internal/modules/cjs/loader:1149:10)\n' +
	'    at Object.require.extensions.<computed> [as .ts] (C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\ts-node\\src\\index.ts:1374:12)\n' +
	'    at Module.load (node:internal/modules/cjs/loader:975:32)\n' +
	'    at Function.Module._load (node:internal/modules/cjs/loader:822:12)\n' +
	'    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)\n' +
	'    at main (C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\ts-node\\src\\bin.ts:331:12)\n' +
	'    at Object.<anonymous> (C:\\Users\\User\\AppData\\Roaming\\npm\\node_modules\\ts-node\\src\\bin.ts:482:3)',
	],
].forEach(([title, e]) =>
{
	test(title, () =>
	{
		let actual = parseErrorStack(e);

		expect(actual).toMatchSnapshot();

		let s = stringifyErrorStack(actual);

		expect(s).toMatchSnapshot();
		expect(s).toStrictEqual(e);
	});

	test(`${title}:02`, () =>
	{
		let actual = parseErrorStack(e, '');

		expect(actual).toMatchSnapshot();

		let s = stringifyErrorStack(actual);

		expect(s).toMatchSnapshot();
		expect(s).toStrictEqual(e);
	});
})





