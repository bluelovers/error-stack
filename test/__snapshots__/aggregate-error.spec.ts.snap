// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty 1`] = `
ErrorStack {
  "code": undefined,
  "message": undefined,
  "rawMessage": "AggregateError",
  "rawStack": "AggregateError
    at Object.<anonymous> (G:\\\\Users\\\\The Project\\\\fork\\\\error-stack\\\\test\\\\temp2.ts:4:11)
    at Module._compile (node:internal/modules/cjs/loader:1097:14)
    at Module.m._compile (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\index.ts:1371:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1149:10)
    at Object.require.extensions.<computed> [as .ts] (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\index.ts:1374:12)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at main (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\bin.ts:331:12)
    at Object.<anonymous> (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\bin.ts:482:3)",
  "rawTrace": Array [
    "    at Object.<anonymous> (G:\\\\Users\\\\The Project\\\\fork\\\\error-stack\\\\test\\\\temp2.ts:4:11)",
    "    at Module._compile (node:internal/modules/cjs/loader:1097:14)",
    "    at Module.m._compile (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\index.ts:1371:23)",
    "    at Module._extensions..js (node:internal/modules/cjs/loader:1149:10)",
    "    at Object.require.extensions.<computed> [as .ts] (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\index.ts:1374:12)",
    "    at Module.load (node:internal/modules/cjs/loader:975:32)",
    "    at Function.Module._load (node:internal/modules/cjs/loader:822:12)",
    "    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)",
    "    at main (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\bin.ts:331:12)",
    "    at Object.<anonymous> (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\bin.ts:482:3)",
  ],
  "traces": Array [
    Object {
      "callee": "Object.<anonymous>",
      "calleeNote": undefined,
      "col": "11",
      "indent": "    ",
      "line": "4",
      "source": "G:\\\\Users\\\\The Project\\\\fork\\\\error-stack\\\\test\\\\temp2.ts",
    },
    Object {
      "callee": "Module._compile",
      "calleeNote": undefined,
      "col": "14",
      "indent": "    ",
      "line": "1097",
      "source": "node:internal/modules/cjs/loader",
    },
    Object {
      "callee": "Module.m._compile",
      "calleeNote": undefined,
      "col": "23",
      "indent": "    ",
      "line": "1371",
      "source": "C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\index.ts",
    },
    Object {
      "callee": "Module._extensions..js",
      "calleeNote": undefined,
      "col": "10",
      "indent": "    ",
      "line": "1149",
      "source": "node:internal/modules/cjs/loader",
    },
    Object {
      "callee": "Object.require.extensions.<computed>",
      "calleeNote": "as .ts",
      "col": "12",
      "indent": "    ",
      "line": "1374",
      "source": "C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\index.ts",
    },
    Object {
      "callee": "Module.load",
      "calleeNote": undefined,
      "col": "32",
      "indent": "    ",
      "line": "975",
      "source": "node:internal/modules/cjs/loader",
    },
    Object {
      "callee": "Function.Module._load",
      "calleeNote": undefined,
      "col": "12",
      "indent": "    ",
      "line": "822",
      "source": "node:internal/modules/cjs/loader",
    },
    Object {
      "callee": "Function.executeUserEntryPoint",
      "calleeNote": "as runMain",
      "col": "12",
      "indent": "    ",
      "line": "77",
      "source": "node:internal/modules/run_main",
    },
    Object {
      "callee": "main",
      "calleeNote": undefined,
      "col": "12",
      "indent": "    ",
      "line": "331",
      "source": "C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\bin.ts",
    },
    Object {
      "callee": "Object.<anonymous>",
      "calleeNote": undefined,
      "col": "3",
      "indent": "    ",
      "line": "482",
      "source": "C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\bin.ts",
    },
  ],
  "type": "AggregateError",
}
`;

exports[`empty 2`] = `
"AggregateError
    at Object.<anonymous> (G:\\\\Users\\\\The Project\\\\fork\\\\error-stack\\\\test\\\\temp2.ts:4:11)
    at Module._compile (node:internal/modules/cjs/loader:1097:14)
    at Module.m._compile (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\index.ts:1371:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1149:10)
    at Object.require.extensions.<computed> [as .ts] (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\index.ts:1374:12)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at main (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\bin.ts:331:12)
    at Object.<anonymous> (C:\\\\Users\\\\User\\\\AppData\\\\Roaming\\\\npm\\\\node_modules\\\\ts-node\\\\src\\\\bin.ts:482:3)"
`;
